ramalama/gpu_detector.py

```diff
@@ -0,0 +1,117 @@
+"""
+MIT License
+
+(C) 2024-2025 ramalama developers
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+"""
+
+import subprocess
+import glob
+
+class GPUDetector:
+    def __init__(self):
+        self.best_gpu = None
+        self.best_vram = 0
+        self.best_env = None
+
+    def _update_best_gpu(self, memory_path, gpu_index, env_var):
+        """
+        Reads GPU memory from a given sysfs path and updates the best available GPU
+        based on the highest VRAM.
+        """
+        try:
+            with open(memory_path, 'r') as file:
+                memory_bytes = int(file.read().strip())
+                memory_mib = memory_bytes // (1024 * 1024)  # Convert bytes to MiB
+
+                if memory_mib > 1024 and memory_mib > self.best_vram:
+                    self.best_vram = memory_mib
+                    self.best_gpu = gpu_index
+                    self.best_env = env_var
+        except Exception as ex:
+            print(f" Warning: Error reading GPU memory info from {memory_path}: {ex}")
+
+    def get_nvidia_gpu(self):
+        """Detects Nvidia GPUs using nvidia-smi and updates the best GPU with the highest VRAM."""
+        print("Trying to detect NVIDIA GPU(s)...")
+        try:
+            result = subprocess.run(
+                ["nvidia-smi", "--query-gpu=index,memory.total", "--format=csv,noheader,nounits"],
+                capture_output=True, text=True, check=True
+            )
+            output = result.stdout.strip()
+
+            for line in output.split('\n'):
+                try:
+                    parts = line.split(',')
+                    if len(parts) != 2:
+                        continue  # Skip malformed lines
+
+                    index, memory_mib = parts[0].strip(), int(parts[1].strip())
+
+                    if memory_mib > self.best_vram:
+                        self.best_vram = memory_mib
+                        self.best_gpu = index
+                        self.best_env = "CUDA_VISIBLE_DEVICES"
+
+                except ValueError as ex:
+                    print(f"Error parsing Nvidia GPU info: {ex}")
+        except FileNotFoundError:
+            print("Warning: `nvidia-smi` not found. No NVIDIA GPU detected or drivers missing.")
```

Comment:
Pleas raise an exception here. And handle this at a higher level.  Warnings should be printed to stderr.